R/scenario_helpers.R

```diff
@@ -0,0 +1,387 @@
+#' Calculate an interval
+#'
+#' @param x A numeric vector.
+#' @param level The interval level as a numeric vector  with values in the range
+#' \eqn{[0, 100]}. Defaults to `95` for the 95% interval.
+#'
+#' @return A numeric vector of the same length as `level`.
+#' @export
+ci <- function(x, level = 95) {
+  checkmate::assert_numeric(x, finite = TRUE, any.missing = FALSE)
+  checkmate::assert_numeric(level, 0, 100, finite = TRUE, any.missing = FALSE)
+
+  z <- 1.0 - (1.0 - (level / 100)) / 2
+
+  stats::qnorm(z) * (mean(x) / sqrt(length(x)))
+}
+
+#' Get pandemic costs from a list of model outputs
+#'
+#' @description See `daedalus::get_costs()` for more information.
+#'
+#' @param l A list of `<daedalus_output>` objects; each object will be passed to
+#' `daedalus::get_costs()`.
+#' @param names A character vector, intended to apply to elements of `l`.
+#'
+#'
+#' @keywords internal
+get_costs_list <- function(l, names) {
+  # no input checks on internal functions
+  costs <- lapply(l, daedalus::get_costs, summarise_as = "domain")
+
+  costs <- do.call(rbind, costs)
+  cost_df <- data.table::as.data.table(costs)
+  cost_df$tag <- names
+
+  cost_df <- data.table::melt(
+    cost_df,
+    id.vars = "tag",
+    variable.name = "domain",
+    value.name = "cost"
+  )
+
+  data.table::setDF(cost_df)
+
+  cost_df
+}
+
+#' Get epidemic summary data from a list of model outputs
+#'
+#' @inheritParams get_costs_list
+#'
+#' @keywords internal
+get_summary_list <- function(l,
+                             names = c("lower", "mean", "upper")) {
+  # no input checks on internal functions
+  summary_data <- Map(
+    l, names,
+    f = function(x, n) {
+      df <- daedalus::get_epidemic_summary(x)
+      df$tag <- n
+
+      df
+    }
+  )
+
+  dt <- data.table::rbindlist(summary_data)
+
+  data.table::setDF(dt)
+}
+
+#' Get incidence data from a list of model outputs
+#'
+#' @inheritParams get_costs_list
+#'
+#' @keywords internal
+get_epidata_list <- function(l, names) {
+  # no input checks on internal functions
+  compartment <- NULL
+  value <- NULL
+  df_list <- lapply(l, daedalus::get_incidence)
+
+  total_hosp_list <- lapply(l, function(x) {
+    z <- daedalus::get_data(x)
+    data.table::setDT(z)
+
+    # NOTE: pass option of summarising by age group
+    z <- z[compartment == "hospitalised",
+      list(value = sum(value)),
+      by = "time"
+    ]
+    z$measure <- "total_hosp"
+
+    z
+  })
+
+  dt_list <- Map(
+    df_list, total_hosp_list, names,
+    f = function(incidence_df, hosp_df, n) {
+      dt <- data.table::rbindlist(
+        list(incidence_df, hosp_df)
+      )
+      dt$tag <- n
+
+      dt
+    }
+  )
+
+  dt <- data.table::rbindlist(dt_list)
+
+  dt
```

Comment:
does this need setDF too?

Reply:
A few of these functions could be merged as well, as they do essentially slightly different things to the same output data.